go/pkg/client/tree.go

```diff
@@ -165,21 +165,85 @@ func getExecRootRelPaths(absPath, execRoot, workingDir, remoteWorkingDir string)
 	return relPath, remoteRelPath, nil
 }
 
+// evalParentSymlinks replaces each parent element in relPath with its target if it's a symlink.
+//
+// Returns the resolved absolute path with a list of parent symlinks (absolute paths) if any.
+// The basename of relPath is not resolved. It remains a symlink if it is one.
+// There are no guaranteed relationship between relPath and the returned path beyond having the same basename.
+// The returned paths may not be clean.
+// Any errors would be from accessing files.
+// Example: execRoot=/a relPath=b/c/d/e.go, b->bb, evaledPath=/a/bb/c/d/e.go, symlinks=[/a/b]
+func evalParentSymlinks(execRoot, relPath string, fmdCache filemetadata.Cache) (evaledPath string, symlinks []string, err error) {
+	b := strings.Builder{}
+	b.WriteString(execRoot)
+	ps := strings.Split(relPath, string(filepath.Separator))
+	lastIndex := len(ps) - 1
+	for i, p := range ps {
+		b.WriteRune(filepath.Separator)
+		if i == lastIndex {
+			// Do not resolve basename.
+			b.WriteString(p)
+			break
+		}
+		path := b.String() + p
+		fmd := fmdCache.Get(path)
+		if fmd.Symlink == nil {
+			// Not a symlink.
+			b.WriteString(p)
+			continue
+		}
+		symlinks = append(symlinks, path)
+		if filepath.IsAbs(fmd.Symlink.Target) {
+			// An absolute target may change the entire path.
+			// E.g. /a/b/c -> /aa/bb/cc
+			b.Reset()
+		}
+		b.WriteString(fmd.Symlink.Target)
+	}
+	return b.String(), symlinks, nil
+}
+
 // loadFiles reads all files specified by the given InputSpec (descending into subdirectories
 // recursively), and loads their contents into the provided map.
 func loadFiles(execRoot, localWorkingDir, remoteWorkingDir string, excl []*command.InputExclusion, filesToProcess []string, fs map[string]*fileSysNode, cache filemetadata.Cache, opts *TreeSymlinkOpts, nodeProperties map[string]*cpb.NodeProperties) error {
 	if opts == nil {
 		opts = DefaultTreeSymlinkOpts()
 	}
 
+	var seenRelPaths map[string]struct{}
+	if opts.Preserved {
+		// Remember loaded files to avoid reprocessing shared symlink ancestors.
+		seenRelPaths = make(map[string]struct{}, len(filesToProcess))
+	}
 	for len(filesToProcess) != 0 {
 		path := filesToProcess[0]
 		filesToProcess = filesToProcess[1:]
 
 		if path == "" {
 			return errors.New("empty Input, use \".\" for entire exec root")
 		}
-		absPath := filepath.Join(execRoot, path)
+		var absPath string
+		if opts.Preserved {
+			// Save here as well to handle the case when a symlink is both a requested file and an ancestor.
+			seenRelPaths[path] = struct{}{}
+			ap, parentSymlinks, err := evalParentSymlinks(execRoot, path, cache)
+			if err != nil {
+				return err
+			}
+			absPath = ap
+			for _, p := range parentSymlinks {
+				if p, err = getRelPath(execRoot, p); err != nil {
```

Comment:
I missed that. Let me unpack it.

The behaviour for leaf symlinks now is basically `Preserved && MaterializeOutsideExecroot == !Preserved` https://github.com/bazelbuild/remote-apis-sdks/blob/6cd9932b11432616a57bf9a6952e7466df90edc2/go/pkg/client/tree.go#L265-L277

To consider an example for an intermediate symlink, let's say `/root/wd/a/b.go` and `a->/foo/bar/aa`, then it should be materialized as `/root/wd/a`. If that is also included as input, it's also going to be materialized, which means there should be no symlink/directory conflict.

Another thing to consider is the evaluated path. In this example, it would be `/foo/bar/aa/b.go`, but it should be `/root/wd/a/b.bo`.

I need to update `evalParentSymlinks` to handle that option while constructing the path.